Helper/Cart.php

```diff
@@ -365,21 +380,37 @@ public function getCartData($payment_only, $place_order_payload)
 				}
 			}
 
-			$cart['shipments'] = @$shipping_address ? [[
-				'cost'             => $this->getRoundAmount($shippingAddress->getShippingAmount()),
-				'tax_amount'       => $this->getRoundAmount($shippingAddress->getShippingTaxAmount()),
-				'shipping_address' => $shipping_address,
-				'service'          => $shippingAddress->getShippingDescription(),
-			]] : null;
+			$cart['shipments'] = [];
+
+
+			if (@$shipping_address) {
+
+				$cost         = $shippingAddress->getShippingAmount();
+				$rounded_cost = $this->getRoundAmount($cost);
+
+				$diff += $cost * 100 - $rounded_cost;
+				$totalAmount += $rounded_cost;
+
+				$cart['shipments'][] = [
+					'cost'             => $rounded_cost,
+					'tax_amount'       => $this->getRoundAmount($shippingAddress->getShippingTaxAmount()),
+					'shipping_address' => $shipping_address,
+					'service'          => $shippingAddress->getShippingDescription(),
+				];
+			}
+
+			$tax_amount         = $shippingAddress->getTaxAmount();
+			$rounded_tax_amount = $this->getRoundAmount($tax_amount);
 
-			$totalAmount = $quote->getGrandTotal();
-			$taxAmount   = $this->getRoundAmount($shippingAddress->getTaxAmount());
+			$diff = $tax_amount * 100 - $rounded_tax_amount;

```

Comment:
can you bugsnag if diff is huge? Also maybe we should error if diff is larger than certain amount (say more than 10 cents).

Reply:
When 2 decimal places are used for price, as it is the case in Magento2, $diff is always 0. This code covers the case that is not likely to happen. Suppose for some reason the 2 decimal price is changed to 4 decimal places. Not straightforward / configurable but pragmatically possible.

For example we have a product with the price, say, 10.0049. and add 30 of them to the cart.
Every single item is rounded to 10 (1000 in Bolt units, cents, whatever) but the real total is 300.147, or 30015 rounded to bolt total. It is 15 cent difference but that is no error. We must add it to the total the way both sides are satisfied. 

While doing this I added random values for additional 2 decimals and added tens of different products to cart to make sure the code works in that case and it did.

One important thing that I found out by testing and I am exploiting here is how the order is verified at Bolt side. The item totals sum (price * qty) is checked against cart total. The item unit price is not considered, moreover you can put any price in that field. I tested by subtracting the price by 1 to 10 and it passed. That implies the total for the item can be changed / increased as well as the cart total as long as the sum of item totals equals the cart total. So if there is a $diff I add it to the cart total and also to the first item total. Item totals are populated with rounded values previously.  

On order creation there are also discounts, shipping and tax taken into account. 

Similarly for shipping and tax call. If there is any truncated value, it is added to the shipping tax.

But, as I said, this covers very unlikely case.

Reply:
bugsnaged